presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.SliceOutput;
+
+import static java.lang.Math.max;
+
+public abstract class BlockEncodingBuffers
+{
+    protected int[] nextBatchOfPositions;
+    protected int nextBatchSize;
+    protected int bufferedRowCount;
+
+    abstract void prepareBuffers();
+
+    abstract void appendLongValues(long[] values);
+
+    abstract void appendNulls(boolean mayHaveNull, boolean[] nulls);
+
+    abstract void appendOffsets(int[] offsets);
+
+    abstract void appendBlock(Block block);
+
+    abstract void writeTo(SliceOutput sliceOutput);
+
+    abstract int getSizeInBytes();
+
+    public void updateBufferedRowCount(int newRowCount)
+    {
+        bufferedRowCount += newRowCount;
+    }
+
+    public void copyPositions(int[] positions, int startIndex, int batchSize)
```

Comment:
Based on the usage of this method, I suggest to rename arguments to offset and length and add a Block argument:

```
    public void copyPositions(int[] positions, int offset, int length, Block block)
    {
        if (nextBatchOfPositions == null || nextBatchOfPositions.length < length) {
            nextBatchOfPositions = new int[length * 2];
        }
        System.arraycopy(positions, offset, this.nextBatchOfPositions, 0, length);
        this.nextBatchSize = length;
        block.appendValuesToBuffers(this);
        bufferedRowCount += length;
    }


    public void appendRows(Page page, int[] positions, int startIndex, int length)
    {
        for (int i = 0; i < page.getChannelCount(); i++) {
            blockEncodingBuffers[i].copyPositions(positions, startIndex, length, page.getBlock(i));
        }
        bufferedRowCount += length;
    }
```